src/torrent.cpp

```diff
@@ -5678,63 +5680,78 @@ namespace {
 		{
-			TORRENT_ASSERT(!j->error);
 			m_file_priority = *p;
 			update_piece_priorities();
 		}
-		else
-		{
-			// in this case, some file priorities failed to get set
-			TORRENT_ASSERT(j->error);
-			TORRENT_ASSERT(!m_share_mode);
 
-			if (alerts().should_post<file_error_alert>())
-				alerts().emplace_alert<file_error_alert>(j->error.ec
-					, resolve_filename(j->error.file), j->error.operation_str(), get_handle());
+		if (!j->error) return;
 
-			set_error(j->error.ec, j->error.file);
-			pause();
-		}
+		// in this case, some file priorities failed to get set
+		TORRENT_ASSERT(!m_share_mode);
+
+		if (alerts().should_post<file_error_alert>())
+			alerts().emplace_alert<file_error_alert>(j->error.ec
+				, resolve_filename(j->error.file), j->error.operation_str(), get_handle());
+
+		set_error(j->error.ec, j->error.file);
+		pause();
 	}
 
 	void torrent::prioritize_files(std::vector<int> const& files)
 	{
 		INVARIANT_CHECK;
 
-		// store file priorities until we get the metadata
-		if (!valid_metadata())
-		{
-			m_file_priority.assign(files.begin(), files.end());
-			return;
-		}
+		if (is_seed()) return;
 
 		std::vector<boost::uint8_t> const new_priority = fix_priorities(files, m_torrent_file->files());
 
-		if (!m_torrent_file->num_pieces() || is_seed() || new_priority == m_file_priority) { return; }
-
 		// storage may be NULL during shutdown
-		if (boost::shared_ptr<piece_manager> const tmp = m_storage)
+		if (m_storage)
 		{
 			inc_refcount("file_priority");
-			m_ses.disk_thread().async_set_file_priority(tmp.get(), new_priority,
-				boost::bind(&torrent::on_file_priority, shared_from_this(), _1));
+			m_ses.disk_thread().async_set_file_priority(m_storage.get()
+				, new_priority, boost::bind(&torrent::on_file_priority, shared_from_this(), _1));
+		}
+		else
+		{
+			m_file_priority = new_priority;
 		}
 	}
 
 	void torrent::set_file_priority(int index, int prio)
 	{
-		if (index < 0 || (valid_metadata() && index >= m_torrent_file->num_files())) { return; }
+		INVARIANT_CHECK;
+
+		if (is_seed()) return;
 
-		// any unallocated slot is assumed to be 4
-		if (int(m_file_priority.size()) <= index && prio == 4) { return; }
+		// setting file priority on a torrent that doesn't have metadata yet is
+		// similar to having passed in file priorities through add_torrent_params.
+		// we store the priorities in m_file_priority until we get the metadata
+		if (index < 0 || (valid_metadata() && index >= m_torrent_file->num_files()))
+		{
+			return;
+		}
 
-		std::vector<int> file_priority(m_file_priority.begin(), m_file_priority.end());        
-		file_priority.resize(std::max(int(file_priority.size()), index + 1), 4);
-		file_priority[index] = prio;        
+		if (prio < 0) prio = 0;
+		else if (prio > 7) prio = 7;
 
-		prioritize_files(file_priority);
+		std::vector<boost::uint8_t> new_priority = m_file_priority;
+		new_priority.resize(std::max(int(new_priority.size()), index + 1), 4);
+		new_priority[index] = prio;
```

Comment:
The same goes here, it might be a bad idea to submit a disk job if nothing has changed.